src/mds/Server.cc

```diff
@@ -6882,6 +6882,13 @@ void Server::handle_client_getvxattr(const MDRequestRef& mdr)
       // since we only handle ceph vxattrs here
       r = -CEPHFS_ENODATA; // no such attribute
     }
+  } if (xattr_name == "ceph.dir.subvolume"sv) {
+    const auto srnode = cur->get_projected_srnode();
+    if (srnode && srnode->is_subvolume()) {
+      *css << "1";
+    } else {
+      r = -CEPHFS_ENODATA; // no such attribute on this inode
```

Comment:
Shouldn't we return `0` instead if the corresponding `srnode` is not marked as subvolume ? Instead of returning `-ENODATA` ?

Reply:
What was the behaviour before the getvxattr change broke this?

Reply:
If the dir is a non-subvolume dir, then returning `0` would imply that the xattr is indeed set and that the value is `"0"` ... wouldn't that be misleading ?

Reply:
> What was the behaviour before the getvxattr change broke this?

need to rebuild an old commit just before the introduction of `getvxattr` and see how things were for two different clients

Reply:
Why not return `0/1` for dirs. and for files will always return `-ENODATA` ?

Reply:
> > What was the behaviour before the getvxattr change broke this?
> 
> need to rebuild an old commit just before the introduction of `getvxattr` and see how things were for two different clients

built old code and tested on vstart cluster
pre-getvxattr code also has the same problem fetching ceph.dir.subvolume
so, technically its not a regression :stuck_out_tongue: 
```
[mchangir@fedora34 build]$ sudo getfattr -n ceph.dir.subvolume /mnt/mycephfs/volumes/g1/sv1
/mnt/mycephfs/volumes/g1/sv1: ceph.dir.subvolume: No such attribute
[mchangir@fedora34 build]$ sudo getfattr -n ceph.dir.subvolume /mnt/mycephfs/volumes/g1/sv1/cc6b8b67-ee46-42f5-993b-9e168a5348c7
/mnt/mycephfs/volumes/g1/sv1/cc6b8b67-ee46-42f5-993b-9e168a5348c7: ceph.dir.subvolume: No such attribute
[mchangir@fedora34 build]$ sudo getfattr -n ceph.dir.subvolume /mnt/mycephfs/volumes/g1
/mnt/mycephfs/volumes/g1: ceph.dir.subvolume: No such attribute
```

Reply:
> Why not return `0/1` for dirs. and for files will always return `-ENODATA` ?

if we are okay with users discovering presence of xattr `ceph.dir.subvolume` with value `0` for all non-subvolume dirs, then I'm okay with this

Reply:
> built old code and tested on vstart cluster
> pre-getvxattr code also has the same problem fetching ceph.dir.subvolume
> so, technically its not a regression

AH. So, now with this change fetching ceph.dir.subvolume is allowed, which I think is fine. As far as the value of the xattr is concerned, IMO its best to keep it simple -- ENODATA if it isn't set, 0/1 if its set with the respective value. Does that sound fine? @mchangir @lxbsz

Reply:
> > built old code and tested on vstart cluster
> > pre-getvxattr code also has the same problem fetching ceph.dir.subvolume
> > so, technically its not a regression
> 
> AH. So, now with this change fetching ceph.dir.subvolume is allowed, which I think is fine. As far as the value of the xattr is concerned, IMO its best to keep it simple -- ENODATA if it isn't set, 0/1 if its set with the respective value. Does that sound fine? @mchangir @lxbsz

Since the `ceph.dir.subvolume` enabling/disabling implementation happens only as a bit twiddle, there's no actual `0` value stored to help identify if the xattr was explicitly set to `0`. So, there's a dilemma of identifying the condition of total absence of the xattr (`ENODATA`) and the xattr value being explicitly set to `0`. Do you see the problem here ?

I don't have a problem with providing an implementation either way. Its the nature of how the current implementation maintains the value of the xattr that's the matter of contention. If we decide to remember the xattr being explicitly being set to `0`, then we need to implement a mechanism to persist the value to disk and test for its presence when replying to a `getvxattr` and return a `"0"` instead of interpreting a bit value of `0` as `ENODATA`

Reply:
We may have to then provide implementation for removing the xattr
i.e. `setfattr -x ceph.dir.subvolume ...`

Reply:
> Why not return `0/1` for dirs. and for files will always return `-ENODATA` ?

This too makes sense 👍

Reply:
ack ... its `0/1` for dirs then ... files are already been taken care of

Reply:
Yeah, sounds good to me.